### DISCUSSION TITLE
Focus an existing Kitty window and a specific tab

### DISCUSSION THREAD

**Discussion by HalibutGitWiz** [kovidgoyal/kitty#8136] [Q&A] [answered]
I usually run a single Kitty window with multiple tabs in it, created by a session (so they are always the same). To help with this, I use [jumpapp](https://github.com/mkropat/jumpapp). I have mapped a key binding to the following command: 

```sh
jumpapp kitty --start-as maximized
```

Pressing the key binding will either start a new maximized Kitty window, or focus the existing one.

That's fine but I'd like to push this further and focus one specific tab in that window. I read [this page](https://sw.kovidgoyal.net/kitty/remote-control/#kitten-focus-tab) and tried modifying my command like this:

```sh
jumpapp kitty -o allow_remote_control=yes -o enabled_layouts=tall --start-as maximized; kitten @ focus-tab --match "id:2"
```

I get the same result as with the first command. Kitty is focused, but on the last tab used and not tab number 2.

I also read #7936 with a similar topic. I'm not sure I understood it well, but I don't think this is the same use case.

Any idea how I could achieve this ?

**Comment by kovidgoyal** [comment]
Use sessions. https://sw.kovidgoyal.net/kitty/overview/#sessions
the example there shows you how to focus specific windows.

**Reply by HalibutGitWiz** [reply]
Thanks for your answer.

I already use sessions to create tabs and windows when opening a new Kitty OS window. I understand I could also focus one particular tab at the same time if I wanted to. But here I'm asking about focusing a particular tab from an **existing** OS window. I don't think sessions are involved in that case.

That's something I used to do with Tmux in my previous terminal emulator. But now that I'm using Kitty, I've removed Tmux from my workflow.

**Reply by kovidgoyal** [reply]
Then use remote control: https://sw.kovidgoyal.net/kitty/remote-control/

**Reply by HalibutGitWiz** [reply]
That's what I tried before asking (see second command in OP) and that didn't work. Maybe I did it wrong. Could you elaborate ?

**Reply by kovidgoyal** [reply, marked as answer]
If you are starting a new kitty instance you use sessions. If you are trying to control an existing kitty instance you use remote control. If you want to control kitty from outside of kitty, then you use remote control with a socket, and send the kitty instance messages over that socket.

**Reply by HalibutGitWiz** [reply]
Thank you very much. I had missed the doc paragraph about using sockets. That makes perfect sense and I was able to achieve exactly what I wanted. I ended up with a hell of a keymap: 
```
zsh -c "jumpapp kitty --start-as maximized -o allow_remote_control=yes --listen-on unix:/tmp/mykitty; kitten @ --to unix:/tmp/mykitty focus-tab --match=id:2"
```

A little breakdown for future readers. What this does is : 

- If no Kitty instance exists, start one, maximized, with remote control enabled, listening on  socket /tmp/mykitty
- If one instance already exists, focus it

Then ask it on its dedicated socket to focus tab 2.

I had to wrap this in `zsh -c` to make sure both commands were fired. Not sure why, this may be specific to my environment.

**Comment by kovidgoyal** [comment]
On Fri, Dec 27, 2024 at 07:39:13AM -0800, Halibut Git Wiz wrote:
> I had to wrap this in `zsh -c` to make sure both commands were fired. Not sure why, this may be specific to my environment.

That's because kitty just runs the command directly not via a shell. If
you want to run multiple commands you have to wrap it in a shell or an
interpreter or whatever.
